### ISSUE TITLE
Explore font stacks

### ISSUE THREAD

**Issue by badlydrawnrob** [badlydrawnrob/anki#12]
## TL;DR

> Currently using [Verdana](https://en.wikipedia.org/wiki/Verdana) and [DejaVu fonts](https://www.1001fonts.com/dejavu-sans-font.html). I might find more elegant fonts later. Would like to use a [variable fonts?](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_fonts/Variable_fonts_guide) when better browser support. Here's the current way to declare [multiple fonts properly](https://www.hacksoft.io/blog/using-multiple-font-files-the-right-way).
>
> - [x] **Views correctly in Safari, Chrome**
> - [x] Requires override on `<body>` for Anki apps.

<img width="759" alt="Screenshot 2023-12-18 at 16 33 32" src="https://github.com/badlydrawnrob/anki/assets/1176875/46ba2a98-c5b9-4297-8aa7-3478e2473575">

Fonts are a bit of an arse to do cross-browser, cross-device, so if you want complete control and consistency use `@font-face` and save your `.ttf` font files to Anki's `collection.media` folder. Loading speed on iOS/Android may be an issue, so probably two custom fonts max.

You'll find `variables/typography.less`, `partials/headings.less` and `themes/cards/...` css files are the places to go to make your edits.

## Good monospaced fonts

- [ ] [Variable fonts?](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_fonts/Variable_fonts_guide)
- [ ] [5 monospaced fonts with ligatures](https://betterwebtype.com/5-monospaced-fonts-with-coding-ligatures/) — FIRA looks nice.
- [ ] **[Menlo](https://en.wikipedia.org/wiki/Menlo_(typeface)) is also another good option** for a monospace font.
- [ ] [Intel One Mono](https://www.intel.com/content/www/us/en/company-overview/one-monospace-font.html)

## Have already checked over these ...

> - **Beware that `@font-face` [should not be nested](https://github.com/less/less.js/issues/4251#event-11203393544)** inside a selector or class.
> - **Font calculations:** Verdana is now our base font, but it's a bit too big by default. The global `font-size: 1rem;` is inherited by, for example, headers so making the font smaller globally will fuck up everything else. For now I'm targeting just the card notes.[^1]

- [x] I think [`system-ui`](https://web.dev/articles/more-variable-font-options-in-chromium-83) will use [San Francisco](https://developer.apple.com/fonts/) on a Mac[^2]
- [x] `ui-monospace` has [poor support](https://caniuse.com/extended-system-fonts) in browsers.
- [x] ⚠️ Changing the default `--font-size` `rem` screws everything else up. Increases size of all fonts, so be careful. Also, `1rem` should allow user-edited `font-size` options in the browser.
- [ ] **Is setting the default `--font-size` with `rem` a bad idea?** ([this article sets up a percentage](https://snook.ca/archives/html_and_css/font-size-with-rem))
- [x] <s>**Does changing `rem` on an element cascade?** (like changing `p` affects `<code>` even if that tag has it's own `rem` declared?</s> — It's safe to change elements with `rem`.
- [x] <s>Page speed time in iOS / Android apps?</s> — Anki's loading speed seems fine so far.
- [ ] Use [base 64 encoding](https://stephenscaff.com/articles/2013/09/font-face-and-base64-data-uri/) for fonts?
- [x] **YES!!** <s>Does the [release deck](https://github.com/badlydrawnrob/anki/releases) import the custom font?</s>
- [x] <s>Edit the `typography.less` and `headings.less`; or `themes/...` files?</s>
    - It seems to be "all of the above" here.
- [x] Using [multiple `@font-face`](https://www.hacksoft.io/blog/using-multiple-font-files-the-right-way) fonts? (seems to work on desktop)

## Using default `--system` font stacks

> **Monospace ([`ui-monospace`](https://qwtel.com/posts/software/the-monospaced-system-ui-css-font-stack/)) fails on Anki (but works in Safari).** Manually writing `'San Francisco Mono'` or `'SF Mono'` is unpredictable in Safari and Anki. System font with `system-ui` seems to work on both.[^3]
> <img width="204" alt="Screenshot 2023-12-06 at 14 48 33" src="https://github.com/badlydrawnrob/anki/assets/1176875/d4ba2a11-0cbd-4b03-a3cb-c3952bfdb235">
> <img width="194" alt="Screenshot 2023-12-06 at 14 48 41" src="https://github.com/badlydrawnrob/anki/assets/1176875/483eec94-dc33-42d5-8ac1-47516f917250">
>
> **Left:** `system-ui` (`SF Mono` on MacOS);  **Right:** [`DejaVu Sans Mono`](https://dejavu-fonts.github.io)
>
> It's a subtle difference (I actually prefer SF Mono), but a good starting point for custom themes. You can see the `L` [has a straight foot](https://www.canva.com/learn/typography-terms/) in SF Mono, but curve in DejaVu.

I feel the Print First CSS `system-ui` font stack is pretty good, but allowing custom fonts is a nice-to-have (and could help with code readability). **However, after playing around with `system-ui` I'm now confused about the best route!**[^2]

- [x] [Child themes](https://github.com/badlydrawnrob/print-first-css/issues/27) — inheriting from `print-first-css` (basic font stack)
- [x] [Fonts in Anki](https://docs.ankiweb.net/templates/styling.html#installing-fonts) must be local files.

## Verdana looks a bit shit at default `--font-size`

<img width="631" alt="before" src="https://github.com/badlydrawnrob/anki/assets/1176875/6ab11d4e-d87a-4844-b460-cb19db8023b3">
<img width="638" alt="after" src="https://github.com/badlydrawnrob/anki/assets/1176875/66c66fc0-1e62-4aa4-918c-5165a5043cef">

## 🤔 Anki custom font quirks

> See the steps in the [Anki documentation](https://docs.ankiweb.net/templates/styling.html#installing-fonts) to add a custom font. The docs say to name your font `myfont` but on desktop I've named it the real font name and seems to work. Add the font to the `collection.media` folder.

### Naming the font properly

> On Anki desktop, using the proper `font-family: 'Font Family Name';` seems to work. However, here **only the second image will work** — use the font name only (`'DejaVu Sans'`) and use `font-weight` and `font-style` to specify the type of font (bold, italic, whatever).

<img width="829" alt="Screenshot 2023-12-07 at 16 20 37" src="https://github.com/badlydrawnrob/anki/assets/1176875/0dfb5a12-2d63-44f1-bd66-3dd365137f88">

<img width="833" alt="Screenshot 2023-12-07 at 16 20 52" src="https://github.com/badlydrawnrob/anki/assets/1176875/c9107670-99cf-4725-96fc-4672fdc0de20">

## ⏳Card loading speed

> <s>You can use a hosted [`<link ...>`](https://www.digitalocean.com/community/tutorials/how-to-load-and-use-custom-fonts-with-css#finding-and-loading-a-font-file-from-a-hosted-service) tag</s> — Anki doesn't expose the `<header>`, so use [`src: url(...)`](https://www.digitalocean.com/community/tutorials/how-to-load-and-use-custom-fonts-with-css#loading-a-self-hosted-font-with-font-face) to load custom fonts from your `collection.media` folder.

There seems to be a potential issue with **[page load speed](https://www.reddit.com/r/Anki/comments/lfckdp/how_do_i_add_a_custom_font_for_mobile_if_the_font/)** when using a custom font on Anki. I also can't find any useful information on using a URL for an `@font-face` with Anki.

[^1]: This is where things get tricky as both `font-size` and `line-height` is dependent on the font type, so if it's a bigger font by default, that's potentially going to fuck up all our hard work with the baseline. Also, it's a bit uncertain where to put those `font-size` changes and there's [some who say that you should use global utility classes](https://adamwathan.me/css-utility-classes-and-separation-of-concerns/) to do so (like Tailwind).
[^2]: I'm [like 70% sure](https://chriscoyier.net/2022/08/02/actually-the-san-francisco-typeface-does-ship-as-a-variable-font/). It might just fallback to `sans-serif`. See below.
[^3]: **Annoying quirks with Anki and Safari:** `'San Francisco Pro'`, or `'SF Pro'` aren't available on the [Font Book app](https://support.apple.com/en-gb/guide/font-book/welcome/mac) but they are available as a [system default](https://css-tricks.com/snippets/css/system-font-stack/). In Safari, `system-ui` seems to pick `'SF Pro'` as it's default, as does `ui-monospace` with `'SF Pro Mono'` — **[`ui-monospace` has poor support](https://caniuse.com/extended-system-fonts) with Anki and browsers!** Manually typing in the `font-family` (using the name of the San Francisco font) doesn't seem to work on Safari anymore, nor with Anki (I think my previous font stack was falling back to `sans-serif`. **`monospace` weirdness:** selects `'Courier'` on Safari and `'Menlo'` on Anki 🤦‍♂️🤬 ... I "think".

**Comment by badlydrawnrob** [comment]
Probably keep this simple with one for Mac/iOS and one for Android. Possibly allow a custom font if using [base 64 encoding](http://base64font.com/)

**Comment by badlydrawnrob** [comment]
[DejaVu Sans](https://www.1001fonts.com/dejavu-sans-font.html) Book looks a lot slicker than DejaVu Sans Bold. It's a bit too clunky and a wider font than book.

**Comment by badlydrawnrob** [comment]
## Renders ok in Safari (codepen)

![IMG_1918](https://github.com/badlydrawnrob/anki/assets/1176875/be95c8fc-025c-41fe-bcb9-1fa721605611)

## Doesn't render properly in Anki app (no idea why)

![IMG_1919](https://github.com/badlydrawnrob/anki/assets/1176875/5c5b02fc-5371-4f1f-822e-f7f88a7de660)

**Comment by badlydrawnrob** [comment]
Fonts are difficult but this is mostly a solved problem. Keep a stripped down explanation of what, exactly you have to do to achieve this again.
